design/design-generic-references.md

```diff
@@ -0,0 +1,197 @@
+# Generic Resource References
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+In many cases, Crossplane managed resources require information that is not
+available when the user creates the resource. One example is that if you'd like to create
+a `Subnet`, you need to supply a VPC ID but if you create the `VPC` at the same
+time then the ID will be available only after the creation, so you need to extract
+the ID and use it in `Subnet`. Today, it's possible to let the provider do this
+operation using cross-resource references. However, this requires knowing the kind of
+the referenced resource and how to extract the value during development time of
+the provider. There are several cases where this is not possible:
+
+* The referencer field may not always target a single kind. For example, a URL
+  field in Route53 could get populated by a URL of an S3 bucket or access endpoint
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources and composite resources will have a top-level spec field
```

Comment:
I'm not 100% sure whether we should do it for composite resources as well. It's useful at that level, too, and it'd make composite resources more similar to managed resources, which is something we want but there are questions to be answered like would we allow claims to be able to edit that? Wouldn't that mean cross-namespace information leak through core crossplane pod? If we won't allow claims to do that, then it's useful for only cases where you want to use composite as part of another composition. Has anyone doing that asked for it? The more I write, I feel inclined to move it to `Future Considerations` section and keep this specific to managed resources, which covers the use cases laid out under background.